### ISSUE TITLE
Asynchronous Lifecycle for Injectables (@PostConstruct, @PreDestroy)

### ISSUE THREAD

**Issue by gagle** [inversify/InversifyJS#551]
Related to #475, https://github.com/inversify/InversifyJS/issues/218#issuecomment-220807400

In the Java world the standard defines the annotations `@PostConstruct` and `@PreDestroy`. This allows you to define lifecycle hooks for the class when it's injected/destroyed by the DI container. This could be translated into ecmascript like this:

```typescript
@injectable()
class A {
  @postConstruct
  protected init(): void {
   console.log('postConstruct A');
  }
  @preDestroy
  protected async destroy(): Promise<void> {
   console.log('preDestroy B');
  }
}

@injectable()
class B {
  public constructor(@inject(A) private _a: A) {}

  @postConstruct
  protected async init(): Promise<void> {
   console.log('postConstruct B');
  }
  @preDestroy
  protected destroy(): void {
   console.log('preDestroy B');
  }
}
```

And this would print when components are resolved:

```typescript
// When dependency tree is resolved, ie. when `container.get(RootConstructor)` is called
postConstruct A
postConstruct B

// When the container is closed, eg. when container.unload() is called
preDestroy B
preDestroy A
```

Special attention to the postConstruct and preDestroy annotations, they can be sync or async.

This currently cannot be done with inversify because the public `get()` signature is synchronous.

`@postConstruct` could be implemented with an async `onActivation()` handler:

```typescript
const partial: BindingInWhenOnSyntax = ...;
partial.onActivation(async (ctx, instance) => {
  // get postConstruct if any
  const postConstructFunctionName = getPostConstruct(instance.constructor);
  if (postConstructFunctionName) {
    // postConstruct exists
    // supose postConstruct is always async
    await instance[fnName].call(instance)
  }
});
```

I know that this is a major change but making all the functions async will provide a lot more flexibility.

**Comment by gagle** [comment]
Note that `@postConstruct` is also useful when dependencies are injected into properties instead of through the constructor. It will be called after they are injected, it's the second chance to init the injectable.

For now, it would be also fine if `@postConstruct` and `@preDestroy` are implemented even if they are synchronous because even JavaEE does not implement them and some workaround must be done (although the result is the same, async init): http://tomee.apache.org/examples-trunk/async-postconstruct/README.html

@remojansen Do you mind if I push a PR with this sync postConstruct and preDestroy annotations?

**Comment by remojansen** [comment]
Hi @gagle, sorry for the late reply. I had some very busy days over the past few weeks...

About `@postConstruct`, that is the way property injection works at the moment. If the instance has not been created by InversifyJS then you can use `@lazyInject`.

About `@preDestroy`,  we could invoke it when `container.unload()` is called but we will only be able to invoke the `@preDestroy method of singleton instances. InversifyJS does not store references to transient references because memory would never be released.

This is a very complex problem in JavaScript because we cannot detect when an object is about to be garbage collected. I still thinking about this, you can read some of my ideas [here](https://gist.github.com/remojansen/d44ed9754e860ba4d7e69c4d285893f4). If I manage to implement a nice solution for custom scopes you will be able to declare:

- `onActivation` equivalent to `@postConstruct`. Invoked after instances are created.
- `onDeactivation`  equivalent to `@preDestroy`. Invoked then a custom scope ends. We will not be able to do it with the default scope because as I said we cannot detect when an object is about to be garbage collected or out of scope.

About, making the API async, I tried once and I ended up with a not very nice dev experience. It forced you to use async / await and not always was ideal. The solution would be to provide two APIs (e.g. `container.get("TYPE")` and `await container.getAsync("TYPE")`) but this will make the library way more complex. I'm not fully convinced this will happen for now.

In summary, for the moment the API will remain sync and instead of implementing `@preDestroy` and `@postConstruct` I will continue thinking about custom scopes because I think it solves this problem and other problems as well. 

I'm closing this issue. The custom scope feature can be tracked at #218.

**Comment by gagle** [comment]
Hi @remojamsen, thanks a lot for the detailed explanation. I get the point with the preDestroy in transient scope, it's not the responsibility of inversify and can be implemented by the lib consumer. However, the postConstruct even if it can be implemented with onActivation, it's preferable to have this function in the class declaration instead of in the composition file. But again, it can be easily implemented by the lib consumer. The main problem here is with the asynchronous nature and if you had problems in the past with this approach then there's nothing to do for now.

**Comment by remojansen** [comment]
@gagle thanks for giving me more details. Kee in mind that the `onActivation` handler is not really designed for class initialization. Initialization makes more sense in the class definition. The real goal of the `onActivation` handler is to allow [Interception](https://msdn.microsoft.com/en-us/library/hh852703(v=pandp.51).aspx):

> Interception is an advanced programming technique that allows you to intercept calls to an object so that you can add additional logic before or after the call. 

If the goal is initialization your class should probably have an `initialize` method instead of using the `onActivation` handler. You can then decide if you want the `initialize` method to be asynchronous or synchronous.

**Comment by gagle** [comment]
Yes, but we need a way to initialize after the class is instantiated. We can create an async `initialize()` method and then in each method call `await this.initialize()`. But this is very ugly. This should be done automatically by the container and as I explained before, this requires the API to be async.

**Comment by igrek8** [comment]
I am curious if it is possible to make thr @postConstruct method to be async. Cause that way I can do some IO right after the instance has been initialized . Any thoughts how to accomplish it?

**Comment by remojansen** [comment]
@crabicode the [class constructor cannot be async](https://stackoverflow.com/questions/36363278/does-async-await-will-allow-us-to-be-used-on-constructors) a decorator could be async if you use some metadata "ready:false" the problem is that then you would have to manually check every n units of time if it is ready because you cannot await a decorator. So this is not possible and it is likely to continue not being possible.

You are going to need to implement your own initialization. Something like:

```ts
@injectable()
class FileReader {

    private _isReady: boolean;
    private _filePath: string;
    private _data: any;
    private _logger: Logger;

    public constructor(logger: Logger) {
        this._logger = logger;
        this._isReady = false;
    }

    public async initialize(filePath: string) {
        return new Promise((res, rej) => {
            fs.open(this._filePath, "r", (err, fd) => {
                if (err) {
                    this._isReady = true;
                    this._logger.info(`File found: ${filePath}`);
                    rej(err);
                } else {
                    this._data = fd;
                    this._isReady = false;
                    this._logger.error(`File error: ${filePath}`);
                    res(this);
                }
            });
        });
    }

    public async read() {
        if (this._isReady) {
            // do something this this._data;
        } else {
            throw new Error();
            this._logger.error(`File error: ${filePath} is not initialized`);
        }
    }

    public async write() {
        if (this._isReady) {
            // do something this this._data;
        } else {
            throw new Error();
            this._logger.error(`File error: ${filePath} is not initialized`);
        }
    }

    public dispose() {
        // close file
    }

}

const fileReader = container.get<FileReader>("FileReader");
const fileReaderReady = await fileReader.initialize();
const text = await fileReaderReady.read();
```

I would try to avoid this kind of class with internal state like "isReady", instead it is better to use a async  factory function. That way the class does not need to keep state. The factory will know what needs to be done and in which order.

**Comment by gagle** [comment]
Async initialization is one of the big things that I'd like to be included. It's weird to have async getter functions to retrieve already initialized data.

**Comment by parisholley** [comment]
@gagle @crabicode check out #1086
